### PR TITLE
Add cloud storage credential management

### PR DESCRIPTION
**Pull request by kthui**
Adds the ability to load cloud storage credentials aggregated into a single JSON file. The [server PR](https://github.com/triton-inference-server/server/pull/4695) provides some documentations on how the file can be used.

### CODE REVIEW

File: src/filesystem.cc [hunk 1]
```diff
@@ -1898,8 +1970,115 @@ S3FileSystem::DeleteDirectory(const std::string& path)
...
+  static void SortCredential(
+      std::vector<std::pair<std::string, FileSystemCredential>>& credential);
+  template <class FileSystemCredential>
+  static Status GetLongestMatchingCredential(
+      const std::vector<std::pair<std::string, FileSystemCredential>>&
+          credentials,
+      const std::string& file_system_prefix, const std::string& path,
+      FileSystemCredential& credential);
+
+  static std::recursive_mutex
```
Comment (rmccorm4) [inline src/filesystem.cc:2005]: Can this be done with a `std::mutex` instead?

File: src/filesystem.cc [hunk 2]
```diff
@@ -1474,21 +1528,39 @@ S3FileSystem::S3FileSystem(
...
         /*useVirtualAdressing*/ false);
   }
 }
 
 S3FileSystem::~S3FileSystem()
 {
-  Aws::ShutdownAPI(options_);
+  delete client_;
+  // shutdown aws API if no more active instance
+  init_mu.lock();
```
Comment (rmccorm4) [inline src/filesystem.cc:1547]: Can use `std::lock_guard` instead

File: src/filesystem.cc [hunk 3]
```diff
@@ -1941,37 +2166,244 @@ GetFileSystem(const std::string& path, FileSystem** file_system)
...
+      case FileSystemType::GCS:
+        return GetFileSystem("gs://", file_system);
+      case FileSystemType::S3:
+        return GetFileSystem("s3://", file_system);
+      case FileSystemType::AS:
+        return GetFileSystem("as://", file_system);
+      default:
+        return Status(
+            Status::Code::UNSUPPORTED,
+            "The requested filesteam cannot be accessed by type");
```
Comment (rmccorm4) [inline src/filesystem.cc:2222]:
```suggestion
            "Unsupported filesystem type");
```

File: src/filesystem.cc [hunk 4]
```diff
@@ -1941,37 +2166,244 @@ GetFileSystem(const std::string& path, FileSystem** file_system)
...
+    // with env var, only LOCAL and GCS are not path-dependent
+    switch (type) {
+      case FileSystemType::LOCAL:
+        return GetFileSystem("", file_system);
+      case FileSystemType::GCS:
+        return GetFileSystem("gs://", file_system);
+      default:
+        return Status(
+            Status::Code::UNSUPPORTED,
+            "The requested filesteam cannot be accessed by type");
```
Comment (rmccorm4) [inline src/filesystem.cc:2234]:
```suggestion
            "Unsupported filesystem type");
```

File: src/filesystem.cc [hunk 5]
```diff
@@ -1941,37 +2166,244 @@ GetFileSystem(const std::string& path, FileSystem** file_system)
...
+    }
+  } else if (status_.Message() == "Use legacy credential") {
+    // with env var, only LOCAL and GCS are not path-dependent
+    switch (type) {
+      case FileSystemType::LOCAL:
+        return GetFileSystem("", file_system);
+      case FileSystemType::GCS:
+        return GetFileSystem("gs://", file_system);
+      default:
```
Comment (rmccorm4) [inline src/filesystem.cc:2231]: Maybe return specific error messages here on S3/Azure for clearer errors

File: src/filesystem.cc [hunk 6]
```diff
@@ -1941,37 +2166,244 @@ GetFileSystem(const std::string& path, FileSystem** file_system)
...
+  }
+  return status_;
+}
+
+Status
+FileSystemManager::ReturnErrorOrReload(
+    const Status& load_status, const Status& error_status,
+    const std::string& path, FileSystem** file_system)
+{
+  if (load_status.Message() == "Cached") {
```
Comment (rmccorm4) [inline src/filesystem.cc:2245]: Maybe make use of `Status::Code::ALREADY_EXISTS` so this is less dependent on specific message wording if changed?

File: src/filesystem.cc (hunk 2 above)
Comment (kthui) [reply src/filesystem.cc:1547]: Yes. Updated.

File: src/filesystem.cc (hunk 1 above)
Comment (kthui) [reply src/filesystem.cc:2005]: `LoadCredential` will call `LoadCredentialFromFile` which both hold the same mutex. It might be easier to use `recursive_mutex` to avoid the same thread from locking on the same mutex twice. 

In terms of making the locking mechanism simple, I think it depends on whether we want to protect the `bool is_credential_cached` variable. Because if credential is not cached, then it is loaded from disk. By protecting this check, we prevents a possible duplicate load from disk from concurrent calls into `LoadCredential` or `LoadCredentialFromDisk`, and force concurrent calls to use cached results. On the other hand, if credential is cached, it terminates right away and release the lock, which should not be any significant reduction in parallism. I am not not sure if there is an easier way to protect this corner case?

File: src/filesystem.cc (hunk 3 above)
Comment (kthui) [reply src/filesystem.cc:2222]: Thanks for catching this. Updated.

File: src/filesystem.cc (hunk 5 above)
Comment (kthui) [reply src/filesystem.cc:2231]: Yes, good suggestion. Updated.

File: src/filesystem.cc (hunk 4 above)
Comment (kthui) [reply src/filesystem.cc:2234]: Updated.

File: src/filesystem.cc (hunk 6 above)
Comment (kthui) [reply src/filesystem.cc:2245]: Yes. Updated.

File: src/filesystem.cc (hunk 1 above)
Comment (rmccorm4) [reply src/filesystem.cc:2005]: It looks like LoadCredentialFromFile is only called directly [here](https://github.com/triton-inference-server/core/blob/84429c6f2cca2181e9a92e3240f54a65bc662b62/src/filesystem.cc#L2248), so maybe you could tweak the logic so `LoadCredential` is the only function ever called, and it calls `LoadCredentialFromFile` internally with the lock already held?

Or something along these lines, there can be a documented expectation that the lock must be held when calling this method, etc.

File: src/filesystem.cc [hunk 7]
```diff
@@ -1898,8 +1970,115 @@ S3FileSystem::DeleteDirectory(const std::string& path)
...
+  template <class FileSystemCredential>
+  static Status GetLongestMatchingCredential(
+      const std::vector<std::pair<std::string, FileSystemCredential>>&
+          credentials,
+      const std::string& file_system_prefix, const std::string& path,
+      FileSystemCredential& credential);
+
+  static std::recursive_mutex
+      mu_;  // for protecting concurrent access into static variables
+  static bool is_credential_cached;
```
Comment (rmccorm4) [inline src/filesystem.cc:2007]: Add an `_` underscore after all class member variable names.

File: src/filesystem.cc (hunk 6 above)
Comment (kthui) [reply src/filesystem.cc:2245]: The `LoadCredential` will only return status SUCCESS, ALREADY_EXISTS or UNAVAILABLE. Except SUCCESS, all dependencies do not return those code, so the code is changed to not depends on any message wording.

File: src/filesystem.cc [hunk 8]
```diff
@@ -70,12 +70,21 @@
...
+#include <mutex>
+
 #include "constants.h"
 #include "status.h"
 #include "triton/common/logging.h"
 
+#define TRITONJSON_STATUSTYPE triton::core::Status
+#define TRITONJSON_STATUSRETURN(M) \
+  return triton::core::Status(triton::core::Status::Code::INTERNAL, (M))
+#define TRITONJSON_STATUSSUCCESS triton::core::Status::Success
```
Comment (rmccorm4) [inline src/filesystem.cc:85]: Are these being used?

File: src/filesystem.cc [hunk 9]
```diff
@@ -2242,7 +2250,7 @@ FileSystemManager::ReturnErrorOrReload(
     const Status& load_status, const Status& error_status,
     const std::string& path, FileSystem** file_system)
 {
-  if (load_status.Message() == "Cached") {
+  if (load_status.ErrorCode() == Status::Code::ALREADY_EXISTS) {
```
Comment (rmccorm4) [inline src/filesystem.cc:2253]: I think these should be `StatusCode()` instead of `ErrorCode()` everywhere

File: src/filesystem.cc [hunk 10]
```diff
@@ -1941,37 +2167,251 @@ GetFileSystem(const std::string& path, FileSystem** file_system)
...
-  switch (type) {
-    case FileSystemType::LOCAL:
-      return GetFileSystem("", file_system);
-    case FileSystemType::GCS:
-      return GetFileSystem("gs://", file_system);
-    default:
-      return Status(
-          Status::Code::UNSUPPORTED,
-          "The requested filesteam can not be accessed by type");
+  const Status& status_ = LoadCredential();
```
Comment (rmccorm4) [inline src/filesystem.cc:2209]: don't use underscore for local variable, use underscores for class variables

```suggestion
  const Status& status = LoadCredential();
```

Comment (kthui) [reply src/filesystem.cc:2209]: Yes. Corrected.

File: src/filesystem.cc (hunk 9 above)
Comment (kthui) [reply src/filesystem.cc:2253]: Yes. Updated.

File: src/filesystem.cc (hunk 7 above)
Comment (kthui) [reply src/filesystem.cc:2006]: Corrected.

File: src/filesystem.cc (hunk 8 above)
Comment (kthui) [reply src/filesystem.cc:85]: Yes, removing them will cause compiler error. This is just how other files import `triton_json.h`.

File: src/filesystem.cc (hunk 1 above)
Comment (kthui) [reply src/filesystem.cc:2004]: I see what you means, grouped them into a single `LoadCredential`.

File: src/filesystem.cc [hunk 11]
```diff
@@ -2001,78 +2000,77 @@ class FileSystemManager {
...
 #ifdef TRITON_ENABLE_S3
-  s3_fs = NULL;
+  s3_fs_ = NULL;
 #endif  // TRITON_ENABLE_S3
 #ifdef TRITON_ENABLE_AZURE_STORAGE
-  as_fs = NULL;
+  as_fs_ = NULL;
 #endif  // TRITON_ENABLE_AZURE_STORAGE
-  local_fs = NULL;
+  local_fs_ = NULL;
```
Comment (rmccorm4) [inline src/filesystem.cc:2055]: Can these be smart pointers to be cleaned up automatically?

File: src/filesystem.cc [hunk 12]
```diff
@@ -2001,78 +2000,77 @@ class FileSystemManager {
...
 #ifdef TRITON_ENABLE_AZURE_STORAGE
-  if (as_fs != NULL)
-    delete as_fs;
+  if (as_fs_ != NULL)
+    delete as_fs_;
 #endif  // TRITON_ENABLE_AZURE_STORAGE
-  if (local_fs != NULL)
-    delete local_fs;
+  if (local_fs_ != NULL)
+    delete local_fs_;
```
Comment (rmccorm4) [inline src/filesystem.cc:2073]: Can these be smart pointers instead to be cleaned up automatically?

File: src/filesystem.cc (hunk 11 above)
Comment (rmccorm4) [inline src/filesystem.cc:2055]: Use `nullptr`

File: src/filesystem.cc [hunk 13]
```diff
@@ -2126,33 +2124,33 @@ FileSystemManager::GetFileSystem(
...
       if (!match_status.IsOk()) {
         return ReturnErrorOrReload(
             load_status, match_status, path, file_system);
       }
-      s3_fs = new S3FileSystem(
+      s3_fs_ = new S3FileSystem(
           path, std::get<0>(credential_).c_str(),
           std::get<1>(credential_).c_str(), std::get<2>(credential_).c_str(),
           std::get<3>(credential_).size() ? std::get<3>(credential_).c_str()
                                           : NULL);
```
Comment (rmccorm4) [inline src/filesystem.cc:2142]: `nullptr` over `NULL`

File: src/filesystem.cc [hunk 14]
```diff
@@ -2242,144 +2240,138 @@ FileSystemManager::GetFileSystem(FileSystemType type, FileSystem** file_system)
...
+  std::lock_guard<std::mutex> lock(mu_);
 
-    triton::common::TritonJson::Value cred_json;
+  // check if credential is already cached
+  if (is_credential_cached_ && !flush_cache) {
+    return Status(Status::Code::ALREADY_EXISTS, "Cached");
+  }
+
+  const char* file_path_c_str = std::getenv("TRITON_CLOUD_CREDENTIAL_PATH");
+  if (file_path_c_str != NULL) {
```
Comment (rmccorm4) [inline src/filesystem.cc:2274]: `nullptr` over `NULL`

File: src/filesystem.cc (hunk 12 above)
Comment (kthui) [reply src/filesystem.cc:2073]: Yes. Switched to `unique_ptr`.

File: src/filesystem.cc (hunk 11 above)
Comment (kthui) [reply src/filesystem.cc:2055]: Switched to `unique_ptr`, no longer applicable.

File: src/filesystem.cc (hunk 13 above)
Comment (kthui) [reply src/filesystem.cc:2142]: Updated.

File: src/filesystem.cc (hunk 14 above)
Comment (kthui) [reply src/filesystem.cc:2274]: Updated.

File: src/filesystem.cc [hunk 15]
```diff
@@ -1898,8 +1969,113 @@ S3FileSystem::DeleteDirectory(const std::string& path)
...
+  if (s3_fs_ != NULL)
+    delete s3_fs_;
+#endif  // TRITON_ENABLE_S3
+#ifdef TRITON_ENABLE_AZURE_STORAGE
+  if (as_fs_ != NULL)
+    delete as_fs_;
+#endif  // TRITON_ENABLE_AZURE_STORAGE
+  if (local_fs_ != NULL)
+    delete local_fs_;
+}
```
[the remark this answers is not in the document]
Comment (kthui) [reply src/filesystem.cc:2013]: Updated. The file system will be cached.

File: src/filesystem.cc [hunk 16]
```diff
@@ -1898,8 +1969,113 @@ S3FileSystem::DeleteDirectory(const std::string& path)
...
+#ifdef TRITON_ENABLE_S3
+std::vector<std::pair<
+    std::string,
+    std::tuple<std::string, std::string, std::string, std::string>>>
+    FileSystemManager::s3_credential_;
+#endif  // TRITON_ENABLE_S3
+#ifdef TRITON_ENABLE_AZURE_STORAGE
+std::vector<std::pair<std::string, std::pair<std::string, std::string>>>
+    FileSystemManager::as_credential_;
+#endif  // TRITON_ENABLE_AZURE_STORAGE
```
[the remark this answers is not in the document]
Comment (kthui) [reply src/filesystem.cc]: Updated. There will only be one file system manager object.

**Comment by kthui** [comment]
There is a greater change in architecture. Previously, the file system manager only handles credentials from file, and let each file system to either initialize from env var or from provided credential.

Currently, the file system manager will handle loading from both credential file and env var, and forms a single credential cache. Therefore, each file system will only initialize from the provided credential, from credential manager. Each credential in cache may also cache its respective file system, specifically, to be cached when the file system is requested.

File: src/filesystem.cc [hunk 17]
```diff
@@ -948,28 +969,29 @@ ASFileSystem::ParsePath(
...
       }
     } else {
-      account_name = std::string(account_str);
+      account_name = account_str;
     }
 
     std::shared_ptr<as::storage_credential> cred;
-    if (account_key != NULL) {
+    if (account_key.size()) {
       // Shared Key
```
Comment (rmccorm4) [inline src/filesystem.cc:995]: 1. I would double check this part, not sure which logic is intended.
```cpp
// Before - these 2 conditions were opposites
if (account_str == NULL)
...
if (account_key != NULL)
...

// After - two conditions are the same?
if (account_str.size())
...
if (account_key.size())
...
```
2. Use `string.empty()` instead of `string.size()` so it's a bit more clear for the reader what the condition checks

File: src/filesystem.cc [hunk 18]
```diff
@@ -1291,10 +1313,13 @@ namespace s3 = Aws::S3;
 
 class S3FileSystem : public FileSystem {
  public:
-  S3FileSystem(const Aws::SDKOptions& options, const std::string& s3_path);
+  S3FileSystem(
+      const std::string& s3_path, const std::string& secret_key,
+      const std::string& key_id, const std::string& region,
+      const std::string& session_token, const std::string& profile_name);
```
Comment (rmccorm4) [inline src/filesystem.cc:1319]: Maybe encapsulate in a simple `S3Options` struct?

File: src/filesystem.cc [hunk 19]
```diff
@@ -1473,22 +1507,46 @@ S3FileSystem::S3FileSystem(
     }
   }
 
-  if ((secret_key != NULL) && (key_id != NULL)) {
-    client_ = s3::S3Client(
+  if (secret_key.size() && key_id.size()) {
```
Comment (rmccorm4) [inline src/filesystem.cc:1510]: `!empty()` instead of `size()`, apply everywhere applicable

File: src/filesystem.cc [hunk 20]
```diff
@@ -1473,22 +1507,46 @@ S3FileSystem::S3FileSystem(
...
+  // FIXME: Calling Aws::ShutdownAPI will result in segmentation fault.
+  //        Currently, the Aws::initAPI will be called at most once at the
+  //        constructor and zero calls to Aws::ShutdownAPI, which is not a
+  //        memory leak and resources will be reclaimed when triton stops.
+  /*
+  std::lock_guard<std::mutex> lock(init_mu_);
+  if (!--init_count_) {
+    Aws::ShutdownAPI(options_);
+  }
+  */
```
Comment (rmccorm4) [inline src/filesystem.cc:1537]: remove if no longer needed

File: src/filesystem.cc [hunk 21]
```diff
@@ -1473,22 +1507,46 @@ S3FileSystem::S3FileSystem(
...
+}
+
+Status
+S3FileSystem::CheckClient(const std::string& s3_path)
+{
+  bool b;
+  if (!IsDirectory(s3_path, &b).IsOk()) {
```
Comment (rmccorm4) [inline src/filesystem.cc:1544]: Do you need to check the value of `b`? At a glance it looks like it's possible to be false even if status is success: https://github.com/triton-inference-server/core/blob/162a4fd7e039bc5904de021922b48b8aa6657bea/src/filesystem.cc#L1568 ?

File: src/filesystem.cc [hunk 22]
```diff
@@ -1941,39 +2110,283 @@ GetFileSystem(const std::string& path, FileSystem** file_system)
...
+          s3_cred_profile_json.AsString(&s3_cred_profile);
+        s3_cache_.push_back(std::make_tuple(
+            s3_cred_name,
+            std::make_tuple(
+                s3_cred_secret_key, s3_cred_key_id, s3_cred_region,
+                s3_cred_session_token, s3_cred_profile),
+            std::shared_ptr<S3FileSystem>()));
+      }
+      SortCache(s3_cache_);
+    }
```
Comment (rmccorm4) [inline src/filesystem.cc:2265]: Can these be moved into internal helpers like `LoadGCSCredentials`, `LoadS3Crendentials`, etc. for brevity?

File: src/filesystem.cc [hunk 23]
```diff
@@ -893,7 +912,9 @@ const std::string AS_URL_PATTERN = "as://([^/]+)/([^/?]+)(?:/([^?]*))?(\\?.*)?";
 
 class ASFileSystem : public FileSystem {
  public:
-  ASFileSystem(const std::string& s3_path);
+  ASFileSystem(
+      const std::string& path, const std::string& account_str,
+      const std::string& account_key);
```
Comment (rmccorm4) [inline src/filesystem.cc:917]: Maybe encapsulate in a simple `AzureStorageOptions` struct?

File: src/filesystem.cc [hunk 24]
```diff
@@ -503,10 +513,19 @@ class GCSFileSystem : public FileSystem {
...
+GCSFileSystem::GCSFileSystem(const std::string& credential_path)
 {
+  static std::string env_name = "GOOGLE_APPLICATION_CREDENTIALS";
+  std::lock_guard<std::mutex> lock(env_mu_);
+#ifdef _WIN32
+  _putenv((env_name + "=" + credential_path).c_str());
+#else
+  setenv(env_name.c_str(), credential_path.c_str(), 1);
+#endif
   client_ = gcs::Client::CreateDefaultClient();
```
Comment (GuanLuo) [inline src/filesystem.cc:529]: There is no way to pass in the credential directly?

File: src/filesystem.cc [hunk 25]
```diff
@@ -948,28 +969,29 @@ ASFileSystem::ParsePath(
...
-  const char* account_str = std::getenv("AZURE_STORAGE_ACCOUNT");
-  const char* account_key = std::getenv("AZURE_STORAGE_KEY");
   std::shared_ptr<as::storage_account> account = nullptr;
   std::string host_name, container, blob_path, query;
   if (RE2::FullMatch(
           path, as_regex_, &host_name, &container, &blob_path, &query)) {
     size_t pos = host_name.rfind(".blob.core.windows.net");
     std::string account_name;
-    if (account_str == NULL) {
+    if (account_str.size()) {
```
Comment (GuanLuo) [inline src/filesystem.cc:983]: Using size() seems wrong, evaluate to "true" if the string has content
```suggestion
    if (account_str.empty()) {
```

File: src/filesystem.cc [hunk 26]
```diff
@@ -948,28 +969,29 @@ ASFileSystem::ParsePath(
...
         account_name = host_name;
       }
     } else {
-      account_name = std::string(account_str);
+      account_name = account_str;
     }
 
     std::shared_ptr<as::storage_credential> cred;
-    if (account_key != NULL) {
+    if (account_key.size()) {
```
Comment (GuanLuo) [inline src/filesystem.cc:994]:
```suggestion
    if (!account_key.empty()) {
```

File: src/filesystem.cc [hunk 27]
```diff
@@ -1322,10 +1347,15 @@ class S3FileSystem : public FileSystem {
   Status ParsePath(
       const std::string& path, std::string* bucket, std::string* object);
   Status CleanPath(const std::string& s3_path, std::string* clean_path);
-  Aws::SDKOptions options_;
-  s3::S3Client client_;
+  const Aws::SDKOptions options_;
+  s3::S3Client* client_;  // init after Aws::InitAPI is called
   re2::RE2 s3_regex_;
+  // enforce order: Aws::InitAPI -> Aws::ShutdownAPI -> Aws::InitAPI ...
```
Comment (GuanLuo) [inline src/filesystem.cc:1353]: Elaborate on why

File: src/filesystem.cc (hunk 17 above)
Comment (kthui) [reply src/filesystem.cc:995]: My bad. Thanks for catching. Corrected.

File: src/filesystem.cc [hunk 28]
```diff
@@ -1427,32 +1457,36 @@ S3FileSystem::CleanPath(const std::string& s3_path, std::string* clean_path)
...
           "s3://(http://|https://|)([0-9a-zA-Z\\-.]+):([0-9]+)/"
           "([0-9a-z.\\-]+)(((/[0-9a-zA-Z.\\-_]+)*)?)")
 {
+  // init aws api if not already
+  init_mu_.lock();
+  if (!init_count_++) {
+    Aws::InitAPI(options_);
+  }
+  init_mu_.unlock();
+
```
Comment (GuanLuo) [inline src/filesystem.cc:1473]:
```suggestion
{
  // init aws api if not already
  std::lock_guard<std::mutex> lk(init_mu_);
  if (!init_count_++) {
    Aws::InitAPI(options_);
  }
}
```

File: src/filesystem.cc (hunk 19 above)
Comment (kthui) [reply src/filesystem.cc:1510]: Updated.

File: src/filesystem.cc [hunk 29]
```diff
@@ -1473,22 +1507,46 @@ S3FileSystem::S3FileSystem(
...
   }
 }
 
 S3FileSystem::~S3FileSystem()
 {
-  Aws::ShutdownAPI(options_);
+  // should delete all client_ before shutdown
+  delete client_;
+  // shutdown aws api if no more S3 instances
+  // FIXME: Calling Aws::ShutdownAPI will result in segmentation fault.
```
Comment (GuanLuo) [inline src/filesystem.cc:1528]: Bug in AWS client? Is there an issue to track that?

File: src/filesystem.cc (hunk 28 above)
Comment (GuanLuo) [reply src/filesystem.cc:1473]: I think we should revert to  `std::call_once`, the Init / Shutdown seems to be a one-time call in respect to the process? Should isolate it from possible multiple S3 clients for different credentials

File: src/filesystem.cc (hunk 20 above)
Comment (kthui) [reply src/filesystem.cc:1537]: Yes. Moved to smart pointer for `client_`, and no more destructors.

File: src/filesystem.cc (hunk 21 above)
Comment (kthui) [reply src/filesystem.cc:1544]: Because this is only to check if the client can connect with S3, it do not matter whether it is a directory or not. Changed the code to check with S3 directly.

File: src/filesystem.cc (hunk 24 above)
Comment (kthui) [reply src/filesystem.cc:554]: Updated.

File: src/filesystem.cc (hunk 18 above)
Comment (kthui) [reply src/filesystem.cc]: Yes. Updated.

File: src/filesystem.cc (hunk 22 above)
Comment (kthui) [reply src/filesystem.cc]: Yes. Updated.

File: src/filesystem.cc (hunk 23 above)
Comment (kthui) [reply src/filesystem.cc]: Updated.

File: src/filesystem.cc (hunk 25 above)
Comment (kthui) [reply src/filesystem.cc]: Yes. Updated.
The mapping from previous char NULL -> empty string, since credential cannot be empty.

File: src/filesystem.cc (hunk 26 above)
Comment (kthui) [reply src/filesystem.cc]: Updated.

File: src/filesystem.cc (hunk 27 above)
Comment (kthui) [reply src/filesystem.cc]: Because AWS requires its S3 client library to be called exactly like: 
```
#include <aws/core/Aws.h>
int main(int argc, char** argv)
{
   Aws::SDKOptions options;
   Aws::InitAPI(options);
   {
      // make your SDK calls here.
   }
   Aws::ShutdownAPI(options);
   return 0;
}
```
[source](https://docs.aws.amazon.com/sdk-for-cpp/v1/developer-guide/basic-use.html)

The `InitAPI` and `ShutdownAPI` are static. They should not be double called and should follow the order.

File: src/filesystem.cc (hunk 28 above)
Comment (kthui) [reply src/filesystem.cc]: Yes. Updated.

File: src/filesystem.cc (hunk 29 above)
Comment (kthui) [reply src/filesystem.cc]: The closest one was in 2019 ([link](https://github.com/aws/aws-sdk-cpp/issues/1263)), but was dismissed because of the calling order (`InitAPI`-> do things -> `ShutdownAPI`).

I suspect this is because the `ShutdownAPI` is called when the program is shutting down, and the order might be slightly different from [the required ordering](https://github.com/triton-inference-server/core/pull/103#discussion_r933497563).

File: src/filesystem.cc (hunk 24 above)
Comment (kthui) [reply src/filesystem.cc:554]: Just found the way to pass credential via argument. Thanks for checking. Updated.

File: src/filesystem.cc [hunk 30]
```diff
@@ -891,10 +935,40 @@ GCSFileSystem::DeleteDirectory(const std::string& path)
...
+  account_key_ = to_str(account_key);
+}
+
+ASCredential::ASCredential(triton::common::TritonJson::Value& cred_json)
+{
+  triton::common::TritonJson::Value account_str_json, account_key_json;
+  if (cred_json.Find("account_str", &account_str_json))
+    account_str_json.AsString(&account_str_);
+  if (cred_json.Find("account_key", &account_key_json))
+    account_key_json.AsString(&account_key_);
```
[the remark this answers is not in the document]
Comment (kthui) [reply src/filesystem.cc:963]: If the JSON contains syntax error, the JSON parser will catch it before reaching this construction steps, and prints for example `error: creating server: Internal - failed to parse the request JSON buffer: Missing a comma or '}' after an object member. at 279`.

If an expected JSON object is of a different type, the JSON parser will not crash and skip forward. Then, credentials cannot be matched and `error: creating server: Not found - Cannot match credential for path  s3://s3-bucket-41498862/models1` is printed for example.

If the JSON is missing some parts of a credential, this is tricky because the Azure construction allows either/both `account_str` and `account_key` to be empty, and will use "default" if it is empty. If the user misses a key or sets it to empty, it might still be valid. The same goes for the S3 construction.

If the credential is invalid, the `check_client()` will catch it, for example with [`log-verbose=1` and S3 `"region": 0`] prints the following:

```
I0802 06:04:24.262804 33401 filesystem.cc:2335] Using credential    for path  s3://s3-bucket-41498862/models1
I0802 06:04:28.612361 33401 metrics.cc:650] Collecting metrics for GPU 0: NVIDIA TITAN V
I0802 06:04:28.612790 33401 tritonserver.cc:2176] ...(skip the table printout)

I0802 06:04:28.612829 33401 server.cc:259] No server context available. Exiting immediately.
error: creating server: Internal - Unable to create S3 filesystem client. Check account credentials.
+ exit 1
```
Basically, accessing the S3 path `s3://s3-bucket-41498862/models1` is using the default/empty S3 credential and gets account credential error, which should signals for checking for the default S3 credential. The only possible issue is the user must remember to set `log-verbose` to print which credential is used for which path.

**Comment by kthui** [comment]
Rebased.

**Review by GuanLuo — DISMISSED** [review]

File: src/filesystem.cc [hunk 31]
```diff
@@ -505,16 +540,17 @@ class GCSFileSystem : public FileSystem {
   google::cloud::StatusOr<gcs::Client> client_;
 };
 
-GCSFileSystem::GCSFileSystem()
+GCSFileSystem::GCSFileSystem(const GCSCredential& gs_cred)
 {
-  client_ = gcs::Client::CreateDefaultClient();
+  auto creds = gcs::oauth2::CreateServiceAccountCredentialsFromJsonFilePath(gs_cred.path_);
+  if (creds) {
```
Comment (rmccorm4) [inline src/filesystem.cc:546]: what happens for invalid creds?

**Review by rmccorm4 — APPROVED** [review]

**Review by rmccorm4 — APPROVED** [review]
Just need some new tests in server PR

**Comment by kthui** [comment]
Rebased.

File: src/filesystem.cc (hunk 31 above)
Comment (kthui) [reply src/filesystem.cc:546]: The `FileSystemManager` will always call `CheckClient` before returning the file system. If the check fails, it reloads the credential and retry. If the check also fails after reload, the error from `CheckClient` is returned and the server gracefully terminates. It will print something like "Unable to create GCS client. Check account credentials."

**Comment by kthui** [comment]
Rebased.

**Comment by rmccorm4** [comment]
What changed since approvals? Hard to tell with force pushes.

**Comment by kthui** [comment]
No change, rebased only.
```
$ git diff e9023eb2316a13397a4cc10b1fb2dafc631cf8ad de6dcd6a9a2bc6192e52e8432a08f30599caab82 --name-only
include/triton/core/tritonserver.h
src/backend_config.cc
src/backend_config.h
src/backend_model.cc
src/backend_model_instance.cc
src/backend_model_instance.h
src/cuda_utils.cc
src/cuda_utils.h
src/model_config_utils.cc
src/pinned_memory_manager.cc
src/response_cache.cc
src/test/response_cache_test.cc
src/tritonserver.cc
src/tritonserver_stub.cc
```
Notice none of the file change involves `filesystem.cc`

**Comment by rmccorm4** [comment]
For future reference, I think you should be able to rebase without force pushing and it won't discard approvals

**Review by rmccorm4 — APPROVED** [review]

**Review by GuanLuo — APPROVED** [review]
